### PR TITLE
Add admission controller for default storage class.

### PR DESCRIPTION
**Pull request by jsafrane**
The admission controller adds a default class to PVCs that do not require any
specific class. This way, users (=PVC authors) do not need to care about
storage classes, administrator can configure a default one and all these PVCs
that do not care about class will get the default one.

The marker of default class is annotation "volume.beta.kubernetes.io/storage-class", which must be set to "true" to work. All other values (or missing annotation) makes the class non-default.

Based on @thockin's code, added tests and made it not to reject a PVC when no class is marked as default.
. 
@kubernetes/sig-storage

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30900)

### CODE REVIEW

**Comment by jsafrane** [comment]
Marking as do-not-merge, I actually did not test if it works, will do that tomorrow. Volunteers are welcome.

**Comment by thockin** [comment]
@erictune @derekwaynecarr  I have no hands-on experience with admission controllers, so I have no idea if this follows established conventions

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// This is a stand-in until we have a real field.  This string should be a const somewhere.
+const classAnnotation = "volume.beta.kubernetes.io/storage-class"
+
+// This indicates that a particular StorageClass nominates itself as the system default.
+const isDefaultAnnotation = "storage-class.beta.kubernetes.io/isDefaultClass"
```
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: should be something like `storageclass.beta.kubernetes.io/is-default-class` - my bad, sorry

**Comment by thockin** [comment]
If we do this we should probably link it into the set of admissions used by default.

**Comment by thockin** [comment]
and will need at least basic e2e

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	PluginName = "SimplePersistentVolumeClaimDefault"
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I would prefer the plug-in have a name that associates it with managing default storage class.  As a result, can we have "StorageClass" somewhere in the name and the plugin pkg?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	stopChan  chan struct{}
+	store     cache.Store
+}
+
+var _ admission.Interface = &claimDefaulterPlugin{}
+
+// newPlugin creates a new admission plugin.
+func newPlugin(kclient clientset.Interface) *claimDefaulterPlugin {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: do other admission plug-ins (that exist or will exist in the future) watch storage classes?

if so, i would prefer we add it to the shared informers so we can avoid additional list/watchers across admission plug-ins in the future.

see what was done for namespaces here to allow multiple admission plug-ins to share a watch on namespaces:

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/namespace/lifecycle/admission.go#L164

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+// This indicates that a particular StorageClass nominates itself as the system default.
+const isDefaultAnnotation = "storage-class.beta.kubernetes.io/isDefaultClass"
+
+// Admit sets the default value of a PersistentVolumeClaim's storage class, in case the user did
+// not provide a value.
+//
+// 1.  Find available StorageClasses.
+// 2.  Figure which is the default
+// 3.  Write to the PVClaim
+func (c *claimDefaulterPlugin) Admit(a admission.Attributes) error {
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: to be clear, you are purposefully doing this logic on both pvc create and update?  do we allow the value to change or be removed by the user on updates?  i am wondering if on update, you need the prevItem to check if that had the default applied and then removed.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+	pvc, ok := a.GetObject().(*api.PersistentVolumeClaim)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	_, found := pvc.Annotations[classAnnotation]
+	if found {
+		// The use asked for a class.
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: nit: s/use/user

is a len check needed here?

Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: or a trim of some kind, i.e. "    " is provided as input.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+// getDefaultClass returns the default StorageClass from the store, or nil.
+func getDefaultClass(store cache.Store) (*extensions.StorageClass, error) {
+	defaultClasses := []*extensions.StorageClass{}
+	for _, c := range store.List() {
+		class, ok := c.(*extensions.StorageClass)
+		if !ok {
+			return nil, errors.NewInternalError(fmt.Errorf("error converting stored object to StorageClass: %v", c))
+		}
+		if class.Annotations[isDefaultAnnotation] == "true" {
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: how many storage classes exist in practice?  should this be an indexer over the store rather than a for ... range every time?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission_test.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+		}
+		attrs := admission.NewAttributesRecord(
+			claim, // new object
+			nil,   // old object
+			api.Kind("PersistentVolumeClaim").WithVersion("version"),
+			claim.Namespace,
+			claim.Name,
+			api.Resource("persistentvolumeclaims").WithVersion("version"),
+			"", // subresource
+			admission.Create,
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission_test.go]: if you want to have the behavior also applied on updates, add a test for update as well.

**Comment by derekwaynecarr** [comment]
Added a number of comments, poke when more review is needed.

**Comment by erictune** [comment]
Suggest not allowing user to change the storage class once PVC is created.

**Comment by jsafrane** [comment]
> Suggest not allowing user to change the storage class once PVC is created.

Ack, however, the class is just an annotation now. I will add a validator to check for this when we have a real PVC.Spec.Class attribute.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 1 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: renamed

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 2 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: IMO it does not really manage StorageClasses, it maniplulates PVCs only, ergo the `persistentvolumeclaim` package.

Renamed to `SimpleDefaultStorageClassForPVC`. There may be different defaulting policies later, this one is called `Simple`.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 3 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: > do other admission plug-ins (that exist or will exist in the future) watch storage classes?

Not sure now, maybe. We'll do shared informer when the second admission plugin arrives.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 4 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: That's a good point. Now, the class is just an annotation and anyone can change it at any time. In future, I think we will have validators to prevent changing pvc.Spec.Class attribute once it's created.

I'm updating the plugin to react to Create events only. If someone deletes the annotation we should not restore it to default. @thockin, opinion?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 5 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: "" is a valid annotation value meaning "I want classless PV", i.e. PV with class "" (as no StorageClass can have Name=="").

And any validation should happen in validators when we have real pvc.Spec.Class attribute, right?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I have no idea how many classes will be there, my personal assumption is 1 class for 90% of deployments and 1-10 for the rest 10%.

**Comment by jsafrane** [comment]
Addressed feedback. TODO:

* e2e test
* make it default

**Comment by jsafrane** [comment]
Tested manually, works as expected.

**Comment by jsafrane** [comment]
@derekwaynecarr, are there any e2e tests for admission controllers? And what if someone installs a cluster without the admission controller that is being tested? I noticed only some `ServiceAccount` tests.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	PluginName = "SimpleDefaultStorageClassForPVC"
```
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I know I am arguing with Derek, but I don't think this is just about PVC.  It just happens that PVC is the only thing attached to storage class today.  If there were some other object that needed a storage class, and the cluster admin wanted to use this simple mode of defaulting, it would be weird to have different defaults for the other (hypothetical) thing.

I think it's a minor point but this name is over-precise.  I won't block on it, but i think it was better before :)

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 3 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: Agree.  For now, it is singleton

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 4 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: Agree, it should be Create only.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I expect a median of 1 - 2 classes and a 99p of 4 - 5.  Wild guess, of course.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	if len(defaultClasses) == 0 {
+		return nil, nil
+	}
+	if len(defaultClasses) > 1 {
+		glog.V(4).Infof("getDefaultClass %s defaults found", len(defaultClasses))
+		return nil, errors.NewInternalError(fmt.Errorf("%d default StorageClasses were found", len(defaultClasses)))
+	}
+	glog.Infof("getDefaultClass one default found")
```
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: not a very useful log.   If you want to log, at least log the name, but better to log both teh PVC name and teh class assigned, which means "do it in the caller"

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 4 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: how do you designate create-but-not-update ?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I am not sure this is correct.

If you are in a multi-zone setup, won't you want 1 default per zone?  Since zone is not configured via StorageClass, it is reasonable that an end-user doesn't know about them at all but admins most certainly want the PVs distributed across zones.

Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: Let me rephrase ...

I would agree with 1-2 classes for most setups *IF* most users are running small clusters in a single zone.

For anyone running multi-zone -- and I don't know if this is the norm -- you'll probably have the same 1-2 classes *per zone*.

This specific thing is coming up right now for discussion downstream in OpenShift Online.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 4 above)
Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: @jsafrane pvc.Spec is currently immutable post Create. Changing the Class won't be allowed with current validation.

**Comment by markturansky** [comment]
We need to discuss multi-zone and how we distribute PVs across zones when end users are not aware of storageClasses.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: requiring per-zone classes is horrible, but it is fallout from inability to
reach a final design in this time window.  But it ONLY matters if the user
is asking for a PVC in a specific zone.

Do we expect most users to specify a zone for their claims or to let the
system spread them?  I expect the latter

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+		if class.Annotations[isDefaultAnnotation] == "true" {
+			defaultClasses = append(defaultClasses, class)
+			glog.V(4).Infof("getDefaultClass added: %s", class.Name)
+		}
+	}
+
+	if len(defaultClasses) == 0 {
+		return nil, nil
+	}
+	if len(defaultClasses) > 1 {
```
Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: could we have many default classes and perhaps this admission controller round-robins between them?

**Comment by thockin** [comment]
We have conflicting requirements.  If a user asks for a PVC and we default
their class we can:

a) assert num_default_classes === 1
b) pick a random default from the set of default classes

I initially argued for (b).  Erin Boyd talked me down.  I can see arguments
in both directions.

If you need to create a per-zone gold, maybe you should also create a
dont-care-about-zone gold and make THAT one the default, and rely on the
in-code spreading

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 10 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: Talk to Erin :)

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: @thockin how would you expect UX to go in Google Container Engine?  Does the user choose a zone for their pod?  If not, how would you distribute their PVCs across zones just as easily as Kube will spread their pods across zones?

Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: A single default storageClass in a multi-zone setup will have the effect of all pods trying to get scheduled into the same zone, making multi-zone useless.

**Comment by markturansky** [comment]
Having our OpenShift Online multi-zone requirements in mind, I would have argued for (b) with you :)

As it is, a single default class won't work for multi-zone, AFAICT. is there something I am missing?

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I expect most users to NOT specify an explicit zone, and if they do, it
will not work with the beta of this feature.  if they don't care which
zone, the PVC will get a PV in a random zone (in which the cluster has
nodes) and the scheduler will respect the PV zone when choosing a node.

I don't think we have a good answer for what happens if a pod asks for
multiple PVCs and they get different zones.  We probably need to cover that
before we can GA this.

Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: the provisioner can pick a random zone.  You can have gold-east, gold-west
and gold.  Make "gold" the default, let it randomize.

**Comment by markturansky** [comment]
Use Case:

Google Container Engine / OpenShift Online both have clusters spanning 10 zones.  Admins want to rely on Kube to spread pods across 10 zones, but EBS/GCEPDs are created within single zones. The scheduler makes sure a pod ends up in the right zone according to its storage.

* I assume we don't want users to have to pick a zone or even know about storage classes.
* Current provisioner requires 1 StorageClass per zone because zone is a configurable option.
* A single default storageClass will put all pods in 1 zone and leave 9 zones devoid of pods.

How do we resolve this and spread PVs across all zones equally?

@smarterclayton @abhgupta

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: > the provisioner can pick a random zone. 

can or does?

Comment (thockin) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: does

Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: I'll need to look through the code again. I didn't see anything like that in the provisioner, but will look deeper.

Comment (markturansky) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: @thockin ok, thanks. I found it in volume/util.go.

So this *does* distribute PVs across all zones and zone should probably *not* be a configured Param in StorageClass?

**Comment by markturansky** [comment]
ok, i missed the bit of code that choses a zone from all the available zones if none is configured in a storageClass :) My concerns were misplaced.

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 6 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: good point, I'll add a note to documentation and examples.

Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: https://github.com/kubernetes/kubernetes/pull/31106

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 9 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: removed

File: plugin/pkg/admission/persistentvolumeclaim/default/admission.go (hunk 8 above)
Comment (jsafrane) [inline plugin/pkg/admission/persistentvolumeclaim/default/admission.go]: So, assuming that this plugin will put default storage class into several hypothetical objects, 

* it should be named just `DefaultStorageClass` (even without `Simple`)
* it should reside in `storageclass` directory, not in `persistentvolumeclaim`.

I'll add an extra commit for this, just in case we'd like to revert back.

**Comment by jsafrane** [comment]
Renamed to `DefaultStorageClass` and moved to `plugin/pkg/admission/storageclass/default/`. No e2e yet (is it even possible?)

File: plugin/pkg/admission/storageclass/default/admission.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+}
+func (a *claimDefaulterPlugin) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// This is a stand-in until we have a real field.  This string should be a const somewhere.
+const classAnnotation = "volume.beta.kubernetes.io/storage-class"
```
Comment (pmorie) [inline plugin/pkg/admission/storageclass/default/admission.go]: isn't this already a const in another package?

File: plugin/pkg/admission/storageclass/default/admission.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+// This indicates that a particular StorageClass nominates itself as the system default.
+const isDefaultAnnotation = "storageclass.beta.kubernetes.io/is-default-class"
```
Comment (pmorie) [inline plugin/pkg/admission/storageclass/default/admission.go]: I think elsewhere we've used `storage.beta.kubernetes.io` -- use that?

**Comment by thockin** [comment]
I am satisfied.  There are a couple nits from other folks.  I'm going to LGTM this and if we lose the race with the bot, we can followup.

+lgtm 

---

Reviewed 23 of 23 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[plugin/pkg/admission/storageclass/default/admission.go, line 109 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30900#-KPTbq-mSHgUJqaf5Fzz-r1-109:-KPo4aFB21bvuYWotLWJ:b-m82kwd) ([raw file](https://github.com/kubernetes/kubernetes/blob/0fb0c07b4cb9c07d30a0d4d27cb4e726db33705a/plugin/pkg/admission/storageclass/default/admission.go#L109)):*
<details><summary><i>Previously, markturansky (Mark Turansky) wrote…</i></summary>
> @jsafrane pvc.Spec is currently immutable post Create. Changing the Class won't be allowed with current validation.
</details>
is this resolved?  I don't see where we specify to do this on create only

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30900#-:-KPo4hUFmBpADoln9nBz:b75lxbw)*

**Comment by derekwaynecarr** [comment]
@jsafrane - can you send a follow-on PR to document this admission controller here:

https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/admin/admission-controllers.md

Thanks!

**Comment by jsafrane** [comment]
@derekwaynecar: filled https://github.com/kubernetes/kubernetes.github.io/pull/1064
